src/src/com/microsoft/aad/adal/IdToken.java

```diff
@@ -23,28 +23,145 @@
+            this.mSubject = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_SUBJECT);
+            this.mTenantId = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_TENANTID);
+            this.mUpn = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_UPN);
+            this.mEmail = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_EMAIL);
+            this.mGivenName = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_GIVEN_NAME);
+            this.mFamilyName = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_FAMILY_NAME);
+            this.mIdentityProvider = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_IDENTITY_PROVIDER);
+            this.mObjectId = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_OBJECT_ID);
+            String expiration = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_PASSWORD_EXPIRATION);
+
+            if (!StringExtensions.IsNullOrBlank(expiration)) {
+                this.mPasswordExpiration = Long.parseLong(expiration);
+            }
+
+            this.mPasswordChangeUrl = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_PASSWORD_CHANGE_URL);
+        }
+    }
+
+    public String getSubject() {
+        return mSubject;
+    }
+
+    public String getTenantId() {
+        return mTenantId;
+    }
+
+    public String getUpn() {
+        return mUpn;
+    }
 
-    String mUpn;
+    public String getGivenName() {
+        return mGivenName;
+    }
+
+    public String getFamilyName() {
+        return mFamilyName;
+    }
 
-    String mGivenName;
+    public String getEmail() {
+        return mEmail;
+    }
+
+    public String getIdentityProvider() {
+        return mIdentityProvider;
+    }
+
+    public String getObjectId() {
+        return mObjectId;
+    }
+
+    public long getPasswordExpiration() {
+        return mPasswordExpiration;
+    }
+
+    public String getPasswordChangeUrl() {
+        return mPasswordChangeUrl;
+    }
+
+    private HashMap<String, String> parseJWT(String idtoken) throws AuthenticationException {
+        String idbody = extractJWTBody(idtoken);
+        // URL_SAFE: Encoder/decoder flag bit to use
+        // "URL and filename safe" variant of Base64
+        // (see RFC 3548 section 4) where - and _ are used in place of +
+        // and /.
+        byte[] data = Base64.decode(idbody, Base64.URL_SAFE);
+        HashMap<String, String> responseItems = new HashMap<String, String>();
+
+        try {
+            String decodedBody = new String(data, "UTF-8");
+            extractJsonObjects(responseItems, decodedBody);
+            return responseItems;
+        } catch (UnsupportedEncodingException exception) {
+            Logger.e(TAG, "The encoding is not supported.", "", ADALError.ENCODING_IS_NOT_SUPPORTED, exception);
+            throw new AuthenticationException(ADALError.ENCODING_IS_NOT_SUPPORTED, exception.getMessage(), exception);
```

Comment:
It's possible if the server sends us bad data for some reason.  I think it is ok for us to wrap errors that are the result of server bugs.  There isn't much that a developer could do with the underlying exception anyway except tell the user something bad happened.